### DISCUSSION TITLE
Problem in Github Action and building latest changes pushed

### DISCUSSION THREAD

**Discussion by aalwaqfi** [community/community#25944] [Actions]
<p>Hi Everyone,</p>
<p>I’m running into an issue where GitHub action/runner (or docker/git) is not using my latest pushed changes or it just reads and builds the Master branch rather than the modified branch. For example, I added an extra print out line but when I  <strong>execute docker container logs </strong> I don’t see the new print out.</p>
<ol>
<li>List item</li>
</ol>
<p>I tried different combinations and so many changes across all files and still didn’t get to work. So, I really would love if someone can point me out to the right direction or the solution.</p>
<p><strong>Setup</strong><br>
The flow is: GitHub Action -&gt; Docker-Compose -&gt; {project}/Dockerfile.<br>
Hosting OS: Linux CentOS<br>
Programming language (if it helps):  .Net Core 3.1/C#<br>
Git version: 2.27.0</p>
<p>My <strong>GitHub Action/workflow</strong> contents</p>
<pre><code class="lang-auto">name: Docker Image CI

on:
  push:
    branches: [ develop ]

jobs:
  build:
    runs-on: self-hosted
    steps:
    - uses: actions/checkout@v2
      with:
        ref: develop
        token: ${{ github.token }}

    - name: Stop &amp; Remove Docker Compose Containers
      run: docker-compose down
    - name: Run Docker Compose
      run: docker-compose up -d

</code></pre>
<p>My <strong>Docker-compose</strong> contents:</p>
<pre><code class="lang-auto">version: '3.8'

services:    
 
    cashier:
        build: 
            context: .
            dockerfile: Cashier/Dockerfile
        container_name: cashier
        ports: 
            - "10000:80"
        networks: 
            - salad-bridge
        restart: always
 
networks: 
    salad-bridge:
        external: true
</code></pre>
<p>My <strong>Dockerfile</strong> contents (Generated by Visual Studio 2019)</p>
<pre><code class="lang-auto">#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.

FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-buster-slim AS base
WORKDIR /app
EXPOSE 80

FROM mcr.microsoft.com/dotnet/core/sdk:3.1-buster AS build
WORKDIR /src
COPY ["Cashier/Cashier.csproj", "Cashier/"]
COPY ["Infrastructure/Infrastructure.csproj", "Infrastructure/"]
RUN dotnet restore "Cashier/Cashier.csproj"
COPY . .
WORKDIR "/src/Cashier"
RUN dotnet build "Cashier.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "Cashier.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "Cashier.dll"]
</code></pre>
<p>If any extra details needed please let me know,</p>
<p>Thanks</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/aalwaqfi">@aalwaqfi</a>,</p>
<p>When you push a new commit to a branch in your repository, if you want this new pushed commit can trigger a workflow to build, test or release the latest source code, you need to do the things below:</p>
<ul>
<li>
<p>Make sure the workflow file is existing on the branch which you push the new commit to (<strong>develop</strong> branch in your case).</p>
</li>
<li>
<p>In the workflow file, when you use the <a href="https://github.com/actions/checkout#checkout-v2" rel="nofollow noopener"><strong>checkout</strong> action</a>, by default the action will fetch only a single commit for the event on the <strong>develop</strong> branch.<br>
For example, you push a new commit “<strong>commit1</strong>” to the <strong>develop</strong> branch, and this new commit trigger the workflow on the <strong>develop</strong> branch. If you set the <strong>checkout</strong> step like as below:</p>
</li>
</ul>
<pre><code class="lang-yaml">- name: Checkout code
  uses: actions/checkout@v2
</code></pre>
<p>the <strong>checkout</strong> action will only fetch “<strong>commit1</strong>” from the <strong>develop</strong> branch by default.<br>
So, if you do not need to checkout other commits from other branches, tags, SHA or repositories, you can keep all the inputs as their default values (just like above example).</p>
<blockquote>
<p><strong>it just reads and builds the Master branch rather than the modified branch.</strong></p>
</blockquote>
<p>If the problem still exists, please make sure you have <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run#enabling-debug-logging" rel="nofollow noopener">enabled debug logging</a> in your repository, and share the repository with us, so that we can check more detailed logs of the workflow runs to analyze the root cause.</p>

**Comment by aalwaqfi** [comment]
<p><a class="mention" href="/brightran">@brightran</a>,</p>
<p>Thank you for the help.</p>
<p>Nothing worked as I’m the only developer that is working on the project (a PoC) and tried to have another runner on my local PC and same issue.</p>
<p>I created a mini repo they reproduce my problem.</p>
<p>Repo Url: <a href="https://github.com/alwaqfi/CashierRepo" rel="nofollow noopener">https://github.com/alwaqfi/CashierRepo</a></p>
<p>I tried 2 runes on the repo</p>
<pre><code class="lang-auto">(Private Repo)
//await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToShortDateString()}");
await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToLongDateString()}");
await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToLongTimeString()}");
await Console.Out.WriteLineAsync($"New Order Number Request {DateTime.Now.Ticks}");

"--&gt; docker container logs cashier

Start Time 07/15/2020
Start Time Wednesday, 15 July 2020
Start Time 14:28:51
New Order Number Request 637304201311515733
</code></pre>
<pre><code class="lang-auto">(Public Repo)
//await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToShortDateString()}");
await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToLongDateString()}");
await Console.Out.WriteLineAsync($"Start Time {DateTime.Now.ToLongTimeString()}");
//await Console.Out.WriteLineAsync($"New Order Number Request {DateTime.Now.Ticks}");

"--&gt; docker container logs cashier"

Start Time 07/15/2020
Start Time Wednesday, 15 July 2020
Start Time 14:32:26
New Order Number Request 637304203460523966
</code></pre>
<p>Thanks</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/aalwaqfi">@aalwaqfi</a>,</p>
<p>From the repository you shared, I do not see the error you reported.<br>
I can see the <strong>checkout</strong> action fetched the right commit from the <strong>develop</strong> branch, not from the <strong>master</strong> branch. Everything looks like normal.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b82cbadd784ccee0b9a965dc2924302122f51a52.png" data-download-href="/uploads/short-url/qhhGEKGLvKGCjoxzyLhaGTCrrjA.png?dl=1" title="example" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181094256-11344fc1-66dd-4fc7-b6a7-bf96b6da408a.png" alt="example" data-base62-sha1="qhhGEKGLvKGCjoxzyLhaGTCrrjA" width="690" height="258" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b82cbadd784ccee0b9a965dc2924302122f51a52_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">example</span><span class="informations">706×265 14.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If I have missed something, please feel free to tell me.</p>

**Comment by aalwaqfi** [comment]
<p><a class="mention" href="/brightran">@brightran</a></p>
<p>Thank you,</p>
<p>The issue is when the action is executed and containers are created but what is in the containers are not my commit contents.</p>
<p>The outputs in my last reply were after the GitHub action completely successfully and containers were created and started. The CI/CD (docker-*) logic seems good, the GitHub action logs seems fine, but what is deployed is not my last commit that the action executed when I pushed it.</p>
<p>I thought something is happening on our side like caching issue. So before I reply, I cloned the repo on my personal machine then executed the action and was able to reproduce the same issue.</p>
<p>Any idea how to proceed?</p>
<p>Note: I’m new to GitHub actions and Docker in general, so the content of the files looks fine to me.</p>
<p>Thanks</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/aalwaqfi">@aalwaqfi</a>,</p>
<blockquote>
<p><strong>I thought something is happening on our side like caching issue.</strong></p>
</blockquote>
<p>Yes, the problem should be caused by the <strong>Docker Cache</strong>.</p>
<p>In the same host machine, by default Docker will cache the results of the first build of a Dockerfile, allowing subsequent builds to be super fast.<br>
However, the cache is used pretty aggressively and may cause issues when you want the updated output of a <strong>RUN</strong> or <strong>ENTRYPOINT</strong> instruction to make it into the new container.  As it stands, unless the Dockerfile itself changes (especially the <strong>RUN</strong> or <strong>ENTRYPOINT</strong> instruction changes, and thus invalidates Docker’s on-host cache), Docker will reuse the previous results from cache. This is clearly disadvantageous when the <strong>RUN</strong> command is a source code checkout, for example a git clone that is <strong>RUN</strong> as the initial step of a project’s build.</p>
<p>To avoid the issues caused by the <strong>Docker Cache</strong>, you can try the following options of the ‘<strong>docker-compose up</strong>’ command (more details, see <a href="https://docs.docker.com/compose/reference/up/" rel="nofollow noopener">here</a>):</p>
<div class="md-table">
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>–force-recreate</td>
<td>Recreate containers even if their configuration and image haven’t changed.</td>
</tr>
<tr>
<td>–build</td>
<td>Build images before starting containers.</td>
</tr>
</tbody>
</table>
</div>

**Comment by aalwaqfi** [comment]
<p><a class="mention" href="/brightran">@brightran</a></p>
<p>Thank you so much,</p>
<p>The  –force-recreate didn’t work. Because I already executing docker-compose down then up, so this should recreate the containers.</p>
<p>The issue seems is in the cached images. So by Adding a new command to my github action solved this issue. So, my Github action looks like this now:</p>
<pre><code class="lang-auto">    - name: Stop &amp; Remove Docker Compose Containers
      run: docker-compose down
    - name: Run Docker Compose
      run: docker-compose up --build -d
</code></pre>
<p>Thanks again,</p>
